src/render/renderer.cpp

```diff
@@ -188,12 +188,46 @@ float Renderer::bisectionAccuracy(const Ray& ray, float t0, float t1, float isoV
     return 0.0f;
 }
 
-// ======= TODO: IMPLEMENT ========
 // In this function, implement 1D transfer function raycasting.
 // Use getTFValue to compute the color for a given volume value according to the 1D transfer function.
 glm::vec4 Renderer::traceRayComposite(const Ray& ray, float sampleStep) const
 {
-    return glm::vec4(0.0f);
+    // return this->frontToBackCompositing(ray, sampleStep);
+    return this->backToFrontComposite(ray, sampleStep);
+}
+
+glm::vec4 Renderer::backToFrontComposite(const Ray& ray, float sampleStep) const
+{
+    glm::vec3 samplePos = ray.origin + ray.tmax * ray.direction;
+    const glm::vec3 increment = sampleStep * ray.direction;
+
+    glm::vec3 color(0.0f);
+
+    for (float t = ray.tmax; t >= ray.tmin; t -= sampleStep, samplePos -= increment) {
+        glm::vec4 tf_value = this->getTFValue(this->m_pVolume->getVoxelInterpolate(samplePos));
+        color = tf_value[3] * glm::vec3(tf_value) + (1 - tf_value[3]) * color;
+    }
+
+    return glm::vec4(color, 1);
+}
+
+// TODO check why this does not work as back to front
+glm::vec4 Renderer::frontToBackCompositing(const Ray& ray, float sampleStep) const
+{
+    glm::vec3 samplePos = ray.origin + ray.tmin * ray.direction;
+    const glm::vec3 increment = sampleStep * ray.direction;
+
+    glm::vec3 color(1.0f);
+    float opacity = 1.0f;
+
+    for (float t = ray.tmin; t <= ray.tmax; t += sampleStep, samplePos += increment) {
+        glm::vec tf_value = this->getTFValue(this->m_pVolume->getVoxelInterpolate(samplePos));
+
+        color = opacity * color + (1 - opacity) * glm::vec3(tf_value);
+        opacity = opacity + (1 - opacity) * tf_value[3];
+    }
+
+    return glm::vec4(color / m_pVolume->maximum(), 1);

```

Comment:
Same issue, doesn't work for back to front compositing with the same logic, but works when the loop is reversed (front to back)